### PR TITLE
[PINT-1852a]

### PR DESCRIPTION
**Pull request by jumpingGrendel**
Adding an Action that allows for registering an email address with Airship and associating it with a Named User ID. Also allows for replacing an email address with a new one.

## Testing

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

**Comment by joe-ayoub-segment** [comment]
Thanks for the PR @jumpingGrendel - I'll schedule for review.

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      if (old_email_response_content.channel.channel_id) {
+        return await register(request, settings, payload, old_email_response_content.channel.channel_id)
+      }
+    }
+    const register_response = await register(request, settings, payload, null)
+    const response_content = register_response.content
+    const data = JSON.parse(response_content)
+
+    if (!data.ok || !data.channel_id) {
+      throw new Error(`Registration Failed, didn't create channel_id for ${payload.channel_object.address}`)
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:159]: We can't allow regular JS Error objects to be thrown. We need to use one of the error types defined in the framework. See docs https://github.com/segmentio/action-destinations/blob/main/docs/error-handling.md 

In this case will we be throwing a non retryable error? If so then I think you should throw a IntegrationError. 

Make sure you provide an error message, an error code, and an error status number. For example: 

   ```
 throw new IntegrationError(
      'Empty Salesforce instance URL. Please login through OAuth.',
      'INVALID_INSTANCE_URL',
      400
    )
```

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    const channel_id = data.channel_id
+    if (payload.named_user_id && payload.named_user_id.length > 0) {
+      const associate_response = await associate_named_user(request, settings, channel_id, payload.named_user_id)
+      if (!associate_response.ok) {
+        throw new Error(`Associate Failed for named user ${payload.named_user_id}`)
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:165]: Same as above - we can't throw a regular JS Error object.

File: packages/destination-actions/src/destinations/airship/utilities.ts [hunk 3]
```diff
@@ -3,6 +3,110 @@ import type { Settings } from './generated-types'
...
+  old_channel: string | null
+) {
+  let address_to_use = payload.channel_object.address
+  const endpoint = map_endpoint(settings.endpoint)
+  let register_uri = `${endpoint}/api/channels/email`
+  if (old_channel && payload.channel_object.new_address) {
+    register_uri = `${endpoint}/api/channels/email/replace/${old_channel}`
+    address_to_use = payload.channel_object.new_address
+  }
+  const country_language = _extract_country_language(payload.locale)
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/airship/utilities.ts:22]: If we change payload.locale to be optional, this code may need to check for null/undefined values.

File: packages/destination-actions/src/destinations/fullstory/generated-types.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 export interface Settings {
   /**
-   * [FullStory Admin API key](https://help.fullstory.com/hc/en-us/articles/360052021773-Managing-API-Keys)
+   * [FullStory API key](https://help.fullstory.com/hc/en-us/articles/360052021773-Managing-API-Keys)
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/fullstory/generated-types.ts:5]: We need to remove this change from the PR. It's for a different Destination.

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi @jumpingGrendel , 
Would you have time to catch up and review this together? I've provided some comments/feedback, but I've some questions about the new Action which would be best discussed in person. 

Could we catch up Thursday or Friday this week?

**Comment by jumpingGrendel** [comment]
> Hi @jumpingGrendel , Would you have time to catch up and review this together? I've provided some comments/feedback, but I've some questions about the new Action which would be best discussed in person.
> 
> Could we catch up Thursday or Friday this week?

Hi @joe-ayoub-segment I can give you 8am Friday (PST) if that works...

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    const response_content = register_response.content
+    const data = JSON.parse(response_content)
+
+    if (!data.ok || !data.channel_id) {
+      throw new IntegrationError(
+        `Registration Failed, didn't create channel_id for ${payload.channel_object.address}`,
+        'Unsuccessful Registration',
+        400
+      )
+    }
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:164]: You might not need to handle errors, as Segment will do it. 
If Airship's API responds with 200 for success and some failure code when a failure happens, there's no need to throw errors yourself.

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  },
+  perform: async (request, { settings, payload }) => {
+    if (payload.channel_object.new_address) {
+      const old_email_channel_response = await getChannelId(request, settings, payload.channel_object.address)
+      const old_email_response_content: any = JSON.parse(old_email_channel_response.content)
+      if (old_email_response_content.channel.channel_id) {
+        return await register(request, settings, payload, old_email_response_content.channel.channel_id)
+      }
+    }
+    const register_response = await register(request, settings, payload, null)
```
[the remark this answers is not in the document]
Comment (jumpingGrendel) [reply packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:154]: In retrospect, I think this update is unneeded. The logic above it (do an email replacement if there's a new email address) actually does `return await` so it won't execute anything  further.

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts (hunk 5 above)
Comment (jumpingGrendel) [reply packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:164]: removed

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        open_tracking_opted_in: { '@path': '$.properties.open_tracking_opted_in' },
+        open_tracking_opted_out: { '@path': '$.properties.open_tracking_opted_out' },
+        transactional_opted_in: { '@path': '$.properties.transactional_opted_in' },
+        transactional_opted_out: { '@path': '$.properties.transactional_opted_out' },
+        suppression_state: { '@path': '$.context.suppression_state' }
+      }
+    }
+  },
+  perform: async (request, { settings, payload }) => {
+    if (payload.channel_object.new_address) {
```
[the remark this answers is not in the document]
Comment (jumpingGrendel) [reply packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:147]: added

File: packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jumpingGrendel) [reply packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:154]: If you review and agree,  I believe  I've  resolved all the  rest of your comments @joe-ayoub-segment . Thanks!

[the remark this answers is not in the document]
Comment (joe-ayoub-segment) [reply packages/destination-actions/src/destinations/airship/registerAndAssociate/index.ts:154]: Hi John, I still think there are a couple of issues with the logic and return types in the perform() function. 
I recorded a [video](https://drive.google.com/file/d/1mIZ0Wwc71FuEB4iL8JHBX5YFxdKgT7I_/view?usp=sharing) to explain. 

Happy to pair on this if you like. Please let me know. 
Joe

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

File: packages/destination-actions/src/destinations/fullstory/generated-types.ts (hunk 4 above)
Comment (jumpingGrendel) [reply packages/destination-actions/src/destinations/fullstory/generated-types.ts]: idk why this file keeps sneaking back in! removed.

File: packages/destination-actions/src/destinations/airship/utilities.ts [hunk 8]
```diff
@@ -3,6 +3,118 @@ import type { Settings } from './generated-types'
...
+      locale_language?: string
+      locale_country?: string
+    }
+  } = {
+    channel: {
+      type: 'email',
+      address: address_to_use
+    }
+  }
+  if (country_language) {
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/airship/utilities.ts:50]:
```suggestion
  if (Array.isArray(country_language) && country_language.length === 2) {
```

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi @jumpingGrendel I think that looks good! makes sense to me. 

One super minor thing and we're done. 

updating country_language check to make it a little more defensive.

**Review by joe-ayoub-segment — APPROVED** [review]

**Comment by jumpingGrendel** [comment]
thanks for the heads-up @joe-ayoub-segment !

**Comment by joe-ayoub-segment** [comment]
Hi @jumpingGrendel this PR has been deployed! Please let me know how everything looks. 
Shall we sync up to do some testing?

**Comment by jumpingGrendel** [comment]
@joe-ayoub-segment I went ahead and put time on your calendar for 4 Sept. Thanks!

**Pull request merged**